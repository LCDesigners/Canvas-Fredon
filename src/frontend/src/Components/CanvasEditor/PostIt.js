import React, { Component } from 'react';
import { TextArea } from 'semantic-ui-react';
import 'semantic-ui-css/semantic.min.css';
import './style.css';
const ReactMarkdown = require('react-markdown');

/**
 * @class PostId : react.component - the plain default PostIt wich holds only text
 */
class PostIt extends Component {

    constructor(...args) {
        super(...args);
        this.textArea = null;
        this.boundPrintListener = this.printListener.bind(this)
    }

    printListener(evt) {
        this.setState({ ...this.state || {}, interactive: !evt.matches })
    }

    UNSAFE_componentWillMount() {
        this.setState({ interactive: true, hover: false, tagbarHeight: null, focus: false })
        window.matchMedia("print").addListener(this.boundPrintListener)
    }

    componentWillUnmount() {
        window.matchMedia("print").removeListener(this.boundPrintListener)
    }

    isEditable() {
        return this.state.interactive && (this.props.editable !== false);
    }

    renderEditor() {

        return (<TextArea
            disabled={!this.isEditable()}
            autoHeight
            rows="1"
            className="textarea"
            value={this.props.model.content.text}
            onChange={this.onChange.bind(this)}
            onKeyDown={this.onKeyDown.bind(this)}
            onFocus={this.onFocus.bind(this)}
            onBlur={this.onBlur.bind(this)}
            ref={(area) => {
                this.textArea = area;
                if ((!area)||(!area.ref)) return;
                if (this.state.tagbarHeight !== area.ref.clientHeight)
                    this.setState({ ...this.state, tagbarHeight: area.ref.clientHeight })
            }}>
        </TextArea>);
    }

    renderViewer() {
        return (<div ref={(viewer) => {
            let clientHeight = (viewer || {}).clientHeight;
            if (clientHeight === undefined) return;//using ```if (!viewer) return``` seems to not catch every "undefined" for some reason (race condition?)
            if (this.state.tagbarHeight !== clientHeight) {
                this.setState({ ...this.state, tagbarHeight: clientHeight })
            }
        }}
            onClick={this.onFocus.bind(this)}>
            <ReactMarkdown //Stateless function component, cannot be used directly with ref (according to react :shrug:)
                autoHeight
                className="markdownnote"
                source={this.props.model.content.text ? this.props.model.content.text : "***"}>
            </ReactMarkdown>
        </div>);
    }

    /**
     * 
     * @param {*} event - the provided event
     * @param {*} data  - the provided event data
     */
    onChange(event, data) {
        let e = this.props.model;
        e.content.text = data.value;
        this.props.callback(e);
    }

    /**
     * 
     * @param {*} event - the provided event 
     */
    onKeyDown(event) {
        if (!this.state.interactive) return;
        this.props.onKeyDown(event, this.props.model);
    }

    onMouseEnter(event) {
        if (!this.state.interactive) return;
        this.setState({ ...this.state || {}, hover: true })
    }

    onMouseLeave(event) {
        this.setState({ ...this.state || {}, hover: false })
    }

    onBlur() {
        this.setState({ ...this.state || {}, focus: false })
    }

    onFocus() {
        if (!this.state.interactive) return;
        this.setState({ ...this.state || {}, focus: true })
    }

    openEntrySettings(evt=null){
        if(evt) evt.stopPropagation()
        this.props.openSettings({ canvas: this.props.canvasID, block: this.props.blockID, entry: this.props.entyID })
    }

    openTags(tagID,evt=null) {
        if(evt) evt.stopPropagation()
        this.props.openSettings({tag:tagID})
    }

    createDragData() {
        let result = {
            source: {
                canvas: this.props.canvasID,
                block: this.props.blockID,
                entry: this.props.entyID
            }
        }
        if (this.textArea) {
            let ref = this.textArea.ref;
            if (ref) {
                let { selectionStart, selectionEnd } = ref;
                if (selectionStart !== selectionEnd)
                    result.textMarkers = { start: selectionStart, end: selectionEnd };
            }
        }
        return result;
    }

    dragStart(event) {
        event.dataTransfer.setData("json", JSON.stringify(this.createDragData()))
        event.dataTransfer.effectAllowed = "all"
    }

    /**
     * from REACT.Component
     * renders the component
     * - first checks if component is enabled and editable
     */
    render() {
        let center = (!this.props.enableMarkdown) || this.state.hover || this.state.focus ? this.renderEditor() : this.renderViewer();
        let tagbarheight = 20;
        if (this.state.tagbarHeight)
            tagbarheight = this.state.tagbarHeight
        let tags = this.props.model.content.tags || [];
        tags = tags.map((tag) => this.props.globalTags[tag])
        return (<span
            onMouseEnter={this.onMouseEnter.bind(this)}
            onMouseLeave={this.onMouseLeave.bind(this)}
            onDragOver={this.props.onDragOver}
            onDrop={this.props.onDrop}
            draggable="true"
            title={this.props.model.content.title}
            onDragStart={this.dragStart.bind(this)}
        >
            <span className="tagbar">
                <svg width="8" height={tagbarheight} preserveAspectRatio="xMinYMin" onClick={this.state.interactive && this.openEntrySettings.bind(this)} cursor={this.state.interactive && "pointer"}>
                    {tags.length?tags.map((tag, i) => (
                        <rect
                            height={tagbarheight / tags.length}
                            y={(tagbarheight / tags.length) * i}
                            width="8"
                            key={tag._id}
                            onClick={this.state.interactive && ((evt)=>evt.shiftKey?this.openTags(tag._id,evt):this.openEntrySettings(evt))}
                            fill={tag.color} />)):undefined
                        }
                </svg>
            </span>
            {center}
        </span>);
    }

}

export default PostIt;
