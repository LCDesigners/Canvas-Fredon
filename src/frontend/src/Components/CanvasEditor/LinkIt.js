/* eslint-disable*/
import React from 'react';
import PostIt from './PostIt'
import {TextArea, Container, Icon} from 'semantic-ui-react';
import 'semantic-ui-css/semantic.min.css';
import './style.css'

import { Link } from 'react-router-dom';

/**
 * @class LinkIt : react component - kind of PostIt wich holds a link an a reference
 */
class LinkIt extends PostIt{

  /**
   * @deprecated
   */
  onReferenceCall(){
    this.props.onReferenceCall(this.props.model);
  }

  /**
   * from REACT.Component
   * renders the component
   * - first checks if component is enabled and editable
   * - provides a link wich leads to the connected canvas
   */
  render() {

    //if editor is in HISTORY mode, the link should lead to its historical reference
    let searchString = "?id=" + this.props.model.content.reference;
    if(this.props.mode === 'HISTORY'){
      searchString += '&tag=' + this.props.tagID;
    }

    return (
      <div className='x-postit' style={{'borderColor': this.props.color}}>
        {super.render()}
        {!this.state.interactive?"":
        <Container onClick={onForceReload}>
        <Link to={{
                    pathname: "/projects/" + this.props.project_id + "/editor/"
                        + "valueproposition",
                    search: searchString
                  }}
                  >
                  <Container>
                    <Icon name="window maximize"></Icon>open
                  </Container>
        </Link>
        </Container>}
       </div>
    );
  }
}

function onForceReload(){
  window.location.reload();
}

export default LinkIt
