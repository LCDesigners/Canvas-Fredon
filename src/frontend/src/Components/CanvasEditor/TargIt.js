import React from 'react';
import PostIt from './PostIt';
import 'semantic-ui-css/semantic.min.css';
import './style.css'

/**
 * @class LinkIt : react component - kind of PostIt wich holds a link an a reference
 */
class TargIt extends PostIt {

  /**
   * from REACT.Component
   * renders the component
   * - first checks if component is enabled and editable
   */
  render() {
    return (
      <div className="x-postit" style={{'borderColor': this.props.color}}>
        {super.render()}
      </div>
    );
  }

}

export default TargIt;
