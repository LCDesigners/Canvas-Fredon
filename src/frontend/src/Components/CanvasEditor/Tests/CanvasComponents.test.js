import React from 'react';
import renderer from 'react-test-renderer';

import LinkIt from '../LinkIt';
import PostIt from '../PostIt';
import TargIt from '../TargIt';

// mock for matchMedia from https://stackoverflow.com/questions/39830580/jest-test-fails-typeerror-window-matchmedia-is-not-a-function
window.matchMedia = window.matchMedia || function() {
  return {
    matches : false,
    addListener : function() {},
    removeListener: function() {}
  };
};


test('PostIt should display the expected state', () => {
    let entry = {
        _id: 'id',
        title: 'title',
        content : {
            text : 'text'
        }
    }
    let editable = true;

    let component = renderer.create(
      <PostIt 
        title = {entry.title}
        text = {entry.content.text}
        entyID = {entry._id} 
        model = {entry} 
        callback =  {undefined} 
        editable= {editable}
        >
      </PostIt>,
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    editable = false;
    component = renderer.create(
        <PostIt 
          title = {entry.title}
          text = {entry.content.text}
          entyID = {entry._id} 
          model = {entry} 
          callback =  {undefined} 
          editable= {editable}
          >
        </PostIt>,
    );

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    entry.content.text = 'other text';

    editable = false;
    component = renderer.create(
        <PostIt 
          title = {entry.title}
          text = {entry.content.text}
          entyID = {entry._id} 
          model = {entry} 
          callback =  {undefined} 
          editable= {editable}
          >
        </PostIt>,
    );

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('TargIt should display the expected state', () => {
    let entry = {
        _id: 'id',
        title: 'title',
        content : {
            text : 'text'
        }
    }
    let editable = true;
    let color = '#ffcc66';

    let component = renderer.create(
      <TargIt 
        title = {entry.title}
        text = {entry.content.text}
        entyID = {entry._id} 
        model = {entry} 
        callback =  {undefined} 
        editable= {editable}
        color= {color}
        >
      </TargIt>,
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    editable = false;
    color = '#ffaaaa';

    component = renderer.create(
        <TargIt 
          title = {entry.title}
          text = {entry.content.text}
          entyID = {entry._id} 
          model = {entry} 
          callback =  {undefined} 
          editable= {editable}
          color= {color}
          >
        </TargIt>,
    );

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    entry.content.text = 'other text';
    editable = true;
    color = '#00ff4a';

    component = renderer.create(
        <TargIt 
          title = {entry.title}
          text = {entry.content.text}
          entyID = {entry._id} 
          model = {entry} 
          callback =  {undefined} 
          editable= {editable}
          color= {color}
          >
        </TargIt>,
    );

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
