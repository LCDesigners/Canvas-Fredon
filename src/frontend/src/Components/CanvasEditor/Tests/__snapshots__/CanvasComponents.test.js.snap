// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostIt should display the expected state 1`] = `
<span
  draggable="true"
  onDragStart={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <span
    className="tagbar"
  >
    <svg
      cursor="pointer"
      height={20}
      onClick={[Function]}
      preserveAspectRatio="xMinYMin"
      width="8"
    />
  </span>
  <textarea
    className="textarea"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    rows="1"
    style={
      Object {
        "resize": "none",
      }
    }
    value="text"
  />
</span>
`;

exports[`PostIt should display the expected state 2`] = `
<span
  draggable="true"
  onDragStart={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <span
    className="tagbar"
  >
    <svg
      cursor="pointer"
      height={20}
      onClick={[Function]}
      preserveAspectRatio="xMinYMin"
      width="8"
    />
  </span>
  <textarea
    className="textarea"
    disabled={true}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    rows="1"
    style={
      Object {
        "resize": "none",
      }
    }
    value="text"
  />
</span>
`;

exports[`PostIt should display the expected state 3`] = `
<span
  draggable="true"
  onDragStart={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <span
    className="tagbar"
  >
    <svg
      cursor="pointer"
      height={20}
      onClick={[Function]}
      preserveAspectRatio="xMinYMin"
      width="8"
    />
  </span>
  <textarea
    className="textarea"
    disabled={true}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    rows="1"
    style={
      Object {
        "resize": "none",
      }
    }
    value="other text"
  />
</span>
`;

exports[`TargIt should display the expected state 1`] = `
<div
  className="x-postit"
  style={
    Object {
      "borderColor": "#ffcc66",
    }
  }
>
  <span
    draggable="true"
    onDragStart={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="tagbar"
    >
      <svg
        cursor="pointer"
        height={20}
        onClick={[Function]}
        preserveAspectRatio="xMinYMin"
        width="8"
      />
    </span>
    <textarea
      className="textarea"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onInput={[Function]}
      onKeyDown={[Function]}
      rows="1"
      style={
        Object {
          "resize": "none",
        }
      }
      value="text"
    />
  </span>
</div>
`;

exports[`TargIt should display the expected state 2`] = `
<div
  className="x-postit"
  style={
    Object {
      "borderColor": "#ffaaaa",
    }
  }
>
  <span
    draggable="true"
    onDragStart={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="tagbar"
    >
      <svg
        cursor="pointer"
        height={20}
        onClick={[Function]}
        preserveAspectRatio="xMinYMin"
        width="8"
      />
    </span>
    <textarea
      className="textarea"
      disabled={true}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onInput={[Function]}
      onKeyDown={[Function]}
      rows="1"
      style={
        Object {
          "resize": "none",
        }
      }
      value="text"
    />
  </span>
</div>
`;

exports[`TargIt should display the expected state 3`] = `
<div
  className="x-postit"
  style={
    Object {
      "borderColor": "#00ff4a",
    }
  }
>
  <span
    draggable="true"
    onDragStart={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="tagbar"
    >
      <svg
        cursor="pointer"
        height={20}
        onClick={[Function]}
        preserveAspectRatio="xMinYMin"
        width="8"
      />
    </span>
    <textarea
      className="textarea"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onInput={[Function]}
      onKeyDown={[Function]}
      rows="1"
      style={
        Object {
          "resize": "none",
        }
      }
      value="other text"
    />
  </span>
</div>
`;
